### ISSUE TITLE
[BUG] __file__ is not the path of pyx file, it points to compiled shared library instead.

### ISSUE THREAD

**Issue by evan0greenup** [cython/cython#5453]
### Describe the bug

The `__file__` value in pyx (cython) file is the path to shared library, it is not point to pyx file.

### Code to reproduce the behaviour:

```cython
print(__file__)
```

### Expected behaviour

It should point to .pyx file or .py file.

### Additional context

One possible solution is to manually modify the value of __file__ in Importer and Loader before execute the module.

**Comment by evan0greenup** [comment]
@scoder , What is your opinion, `*.so` file is always located in a temporary directory, it is useless to user.

And `__file__` value are supposed to point at source, instead of intermediate file.

**Comment by scoder** [comment]
What Python and Cython versions have you tried? Does this happen with Cython 3.0?

**Comment by scoder** [comment]
> What is your opinion, `*.so` file is always located in a
> temporary directory, it is useless to user.

Erm, no? It is the file that gets executed. There is actually no guarantee that the original source files that it was created from are available at runtime. They are not needed and must be packaged and installed explicitly by the developer in order to be copied to the runtime environment.


> And `__file__` value are supposed to point at source, instead of
> intermediate file.

I second that. This was corrected in recent Python versions and Cython should behave the same, if possible, because Cython sources are really Python sources.

Note, however, that "__file__" in non-Cython extension modules does not point to the original .c or .cpp file. Those are even more unlikely to be available at runtime.

**Comment by evan0greenup** [comment]
> What is your opinion, `*.so` file is always located in a temporary directory, it is useless to user.
> Erm, no? It is the file that gets executed. There is actually no guarantee that the original source files that it was created from are available at runtime. They are not needed and must be packaged and installed explicitly by the developer in order to be copied to the runtime environment.
> And `__file__` value are supposed to point at source, instead of intermediate file.
> I second that. This was corrected in recent Python versions and Cython should behave the same, if possible, because Cython sources are really Python sources. Note, however, that "__file__" in non-Cython extension modules does not point to the original .c or .cpp file. Those are even more unlikely to be available at runtime.


Source file here means the file user going to be edited. I believe the one of the most important purpose of `__file__` for user is to locate resource file in relative manner.

**Comment by evan0greenup** [comment]
@scoder , I also recommend Cython to preserve pyx file in wheel or other binary package, for the easy of debug in the future. Because text file won't take much disk space. And Cython is extension of Python, user choose to use Python because it is easy to use. For tools like Jetbrains PyCharm, it can provide more type hint and autocomplete and backtrace information based on pyx file in the package.

**Comment by scoder** [comment]
Note that packaging is not done by Cython. We cannot take the decision for you which files you include in your wheels.

**Comment by scoder** [comment]
> one of the most important purpose of `__file__` for user is to locate
> resource file in relative manner. 

For that purpose, pointing to the .so file is probably best, because it is guaranteed to a) exist at runtime and b) be in the right place.

However, I still think we should point to the main source file instead, if available at runtime. And encourage users to include the sources in binary distributions. Not everyone will (want to) do that, though.

**Comment by da-woods** [comment]
> However, I still think we should point to the main source file instead, if available at runtime. And encourage users to include the sources in binary distributions. Not everyone will (want to) do that, though.

I disagree with this, partly because it seems hard to get right reliably and partly because `__file__` is a useful external runtime check that you've actually used a compiled module when a non-compiled source also exists.

The documentation I could find says

> `__file__` is optional (if set, value must be a string). It indicates the pathname of the file from which the module was loaded (if loaded from a file), or the pathname of the shared library file for extension modules loaded dynamically from a shared library

https://docs.python.org/3/reference/import.html

**Comment by da-woods** [comment]
To me, `__file__` pointing at the .so provides useful diagnostic information about what has actually been imported and we shouldn't lose that. If we want to provide information about the original source it shouldn't be at the expense of being accurate about the imported file

**Comment by scoder** [comment]
>> or the pathname of the shared library file for extension modules loaded dynamically from a shared library
>
>https://docs.python.org/3/reference/import.html

Hmm, interesting. I didn't remember that this was documented. However, this wasn't written for the specific case of Cython where the sources are essentially Python sources and often available in a predictable place (next to the binary) at runtime. They are needed in order to show source code in tracebacks, for example.

Let's ask the other way - is there an actual problem with "__file__" pointing to the binary? Something that does not work currently?

**Comment by evan0greenup** [comment]
>Let's ask the other way - is there an actual problem with "__file__" pointing to the binary? Something that does not work currently?

No, but since __file__ is defined, why don't we make it more consistent with Python ecosystem and satisfy user demand? If technically possible.

**Comment by evan0greenup** [comment]
> To me, `__file__` pointing at the .so provides useful diagnostic information about what has actually been imported and we shouldn't lose that. If we want to provide information about the original source it shouldn't be at the expense of being accurate about the imported file

@da-woods 
In my point of will, unlink PyO3 or other ffi stuff, Cython is like a blackbox for write speedup static-typed code to accelerate Python. Even if the document is literally said shared library is OK, but I don't think it is the intent outcome, according to the nature and purpose of Cython.

**Comment by mattip** [comment]
> satisfy user demand

For another data point, I prefer that `__file__` point to the actual file imported. If that is a compiled C-extension, it should point to the shared object. It does not interest me so much if the compiled c-extension came from cython, PyO3, pybind11, or hand-written C-code, I want to know what is being loaded into the interpreter.

**Comment by evan0greenup** [comment]
@scoder 

As far as I see, Cython is something fill the gap between pure Python code and C/C++ or Rust(PyO3) binding toward Python. And the advantage of Cython over PyPy or GraalPython is the runtime support to CPython.

For C/C++ or Rust, the language is just a method for producing shared-library, the users doesn't care about language. Why it is OK to do like that for C/C++ or Rust? Because, they have their own ecosystem and market, being a language to develop Python extension module is just a be a negligible use case. But for Cython, the situation is different, Cython cannot exist without Python. If user want to do some system native project completely get rid of Python. This user certainly will not choose Cython. It will directly use Rust or C/C++ to finish the project.

For Cython, the focus should be targeting on the language itself and interoperability between Cython and Python runtime. For Cython currently it translate to C code, maybe in future to have a more performance gain, it directly translate to LLVM IR code. Or use lower overhead technique other than dynamic loading `*.so` files to keep executing faster. But All those doesn't change the Cython language and how end-user use it. The only difference which user can notice is execution speed.

If Cython want to get interest from new user and new project, beside legacy usage. The points I described above is notable. Otherwise, when Mojo (an AI-inspired general purpose programming language) (https://github.com/modularml/mojo) become open source to public. Cython will likely be deprecated.

**Comment by scoder** [comment]
> I prefer that `__file__` point to the actual file imported

Note that this is not even true for Python. The variable now always points to the .py source file, regardless of whether the .py, .pyc or .pyo file was loaded. The reasoning is that the .py file is the most relevant of the three files for users, the only one that users can actually look into and read. I could argue that the same applies to Cython generated modules, in a way.
